lnrpc+tor: add network connection timeout

yyforyongyu opened the pull request:
This PR enables the user to set up a global connection timeout, also the timeout in connect peer requests.

Previously there is no place to set up a customized connection timeout value, which is default to the tcp timeout set by the OS, often 2 or 3 minutes.

Fixes #4123

carlaKC commented on lnrpc/rpc.proto:
```diff
@@ -947,6 +947,13 @@ message ConnectPeerRequest {
+    requests. Valid uints are "ms", "s", "m", or "h". For example, to set a
+    timeout value of 30 seconds, use "30s".
+    */
+    string timeout = 3;
```
I think it would be more consistent to use `uint64 timeout_seconds` here. Using suffixes is great for cli, but we usually do seconds for rpc.

yyforyongyu replied:
sure, switched.

carlaKC commented on rpcserver.go:
```diff
@@ -1453,6 +1453,29 @@ func (r *rpcServer) ConnectPeer(ctx context.Context,
+		// Set the new timeout value in our net. This will change the timeout
+		// value in our dialer, so we need to restore the system setting once
+		// the call is finished.
+		r.server.cfg.net.SetTimeout(timeout)
```
Is it possible for this to race if we have two connect request with different values at the same time? Eg, request 1 sets value =30s, request 2 sets value = 40s, request 1 completes and replaces with default, request 2 then uses default? 

Another approach would be to add dial options to `Dial` in `tor.Net/Net` interface, so that we can provide these values on a per-call basis.

yyforyongyu replied:
I agree, forgot about checking race condition... Did try with the `Dial` approach though, turned out it required lots of code change. Originally I was thinking only allowing a global timeout so that it's easier to implement. Once that's done I thought, maybe users want to have some connections using a different timeout value, so I took this shortcut as it required minimal code change. Now that there's the race condition, I'll go back to modifying the function `Dial`. Thanks for pointing out!

carlaKC commented on lntest/itest/network.go:
```diff
@@ -0,0 +1,66 @@
+	node *lntest.HarnessNode, req *lnrpc.ConnectPeerRequest) {
+
+	err := wait.Predicate(func() bool {
+		_, err := node.ConnectPeer(ctxt, req)
```
If connect peer blocks until we get a connection (or fail), we don't need the wait predicate here? This is more used for cases where we need multiple calls to wait until come condition is met (eg, call list channels until our channel shows up, or fail).

yyforyongyu replied:
Originally I was trying the test the customized `timeout` was taking effect. Because the connection will fail because of timeout no matter what, I want to make sure it's timing out using the specified value, not the default one. Now the default one uses 120 seconds, which is much smaller than 15 seconds using in the `Predicate`, and the customized value is 1ms. So if the connect peer blocks or fail, it means the customized timeout value is not used, and that's what it's testing.

carlaKC replied:
Still not entirely sure of the use of wait predicate here, since we're using it to once off check a function rather than repeatedly query a condition. But I see what you're going for, and it's probably not worth rolling our own for this specific case.

wpaulino replied:
Hadn't seen this comment before, but using a context timeout on `ConnectPeer` seems like a better approach instead of `wait.Predicate`? If we see a context deadline error, then we know the custom timeout isn't being used.

yyforyongyu commented on cmd/lncli/commands.go:
```diff
@@ -498,13 +498,28 @@ var connectCommand = cli.Command{
+			Name: "timeout",
+			Usage: "The connection timeout value for current request. " +
+				"Valid uints are {ms, s, m, h}.\n" +
+				"                    If not set, the global connection " +
```
fixed

carlaKC commented on watchtower/wtclient/interface.go:
```diff
@@ -95,21 +96,18 @@ type DB interface {
+	dialer tor.DialFunc) (wtserver.Peer, error) {
 
-	return brontide.Dial(localKey, netAddr, dialer)
+	return brontide.Dial(localKey, netAddr, tor.DefaultConnTimeout, dialer)
```
Sometimes we're using the default, and sometimes we're using the value set in config. Is that just so that we don't need to make more interface changes? Seems strange to set a connection timeout in config, but then only have it sometimes apply.

yyforyongyu replied:
I was actually not sure about this change, some suggestions would really be appreciated.
With the timeout being built into the `Dial` function, all the related interfaces now have to be changed to take the new parameter. Currently two services are using the default value, `bootstrapper` and `watchtower`, it was done this way to have least code change, but yeah it is indeed weird... we can of course pass the config value to the services, then,
 - the `DNSSeedBootstrapper` struct would need a new field to store the timeout value.
 - the `watchtower` will change the function `AuthDial` and the type `AuthDialer`, which will cause more changes to the underlying functions.

It was doable, while I feel the original design was better and easier, that we only touch the `tor.net`. And to solve the race condition, we make a second dial function, `DialTimeout`, on both `brontide` and `tor.Net` interface. For any requests want a customized timeout value, use the `DialTimeout`. What do you think?

carlaKC replied:
> the DNSSeedBootstrapper struct would need a new field to store the timeout value.

I think this is ok, minimal change that lets us thread it through cleanly. 

> authDial and the type AuthDialer, which will cause more changes to the underlying functions

It would be nice to leave these untouched as you say, so we could leave watchtower's `AuthDial` signature unchanged, and then just wrap the call to `brontide.Dial` where we set it in the [server](https://github.com/lightningnetwork/lnd/blob/master/server.go#L1210). Since the existing function is such a thin wrapper already.

yyforyongyu replied:
Sure, the wrapper helps. Fixed according to your suggestions, thanks!

carlaKC commented on tor/net.go:
```diff
@@ -71,13 +84,15 @@ type ProxyNet struct {
 // Dial uses the Tor Dial function in order to establish connections through
 // Tor. Since Tor only supports TCP connections, only TCP networks are allowed.
-func (p *ProxyNet) Dial(network, address string) (net.Conn, error) {
+func (p *ProxyNet) Dial(
```
nit: line wrapping is a bit weird on these, network/address can be on the same line as `Dial`, then newline the rest

yyforyongyu replied:
fixed.

yyforyongyu commented:
> This is looking really good, just a nit on the line wrapping for some of the functions.
> 
> Unfortunately your itests seem to have been sabotaged by some new errors needing whitelist, but a rebase should fix that.

Fixed.

carlaKC commented on server.go:
```diff
@@ -3282,7 +3295,8 @@ type openChanReq struct {
 //
 // NOTE: This function is safe for concurrent access.
-func (s *server) ConnectToPeer(addr *lnwire.NetAddress, perm bool) error {
+func (s *server) ConnectToPeer(
```
Formatting on a few functions is still a bit off, you can leave the first few params in inline then just wrap at 80.

yyforyongyu replied:
changed.

carlaKC commented on tor/net.go:
```diff
@@ -88,7 +103,9 @@ func (p *ProxyNet) LookupHost(host string) ([]string, error) {
 // LookupSRV uses the Tor LookupSRV function in order to resolve SRV DNS queries
 // over Tor.
-func (p *ProxyNet) LookupSRV(service, proto, name string) (string, []*net.SRV, error) {
+func (p *ProxyNet) LookupSRV(
```
These ones too.

yyforyongyu replied:
changed

wpaulino commented on tor/tor.go:
```diff
@@ -125,7 +131,7 @@ func LookupSRV(service, proto, name, socksAddr, dnsServer string,
 
 	// Connect to the DNS server we'll be using to query SRV records.
-	conn, err := dial(dnsServer, socksAddr, streamIsolation)
+	conn, err := dial(dnsServer, socksAddr, streamIsolation, DefaultConnTimeout)
```
`LookupSRV` should also accept the timeout parameter. It's invoked from the `DNSSeedBootstrapper`, which already knows of the configured value.

yyforyongyu replied:
fixed.

wpaulino commented on tor/net.go:
```diff
@@ -42,7 +56,9 @@ func (r *ClearNet) LookupHost(host string) ([]string, error) {
 // LookupSRV for regular network uses net.LookupSRV function
-func (r *ClearNet) LookupSRV(service, proto, name string) (string, []*net.SRV, error) {
+func (r *ClearNet) LookupSRV(service, proto, name string,
+	_ time.Duration) (string, []*net.SRV, error) {
```
Thoughts on also using the timeout here? We can add a context deadline to `net.DefaultResolver.LookupSRV`. which is what `net.LookupSRV` invokes.

yyforyongyu replied:
Sure, wasn't aware of that😂
Timeout context added.

wpaulino commented:
@yyforyongyu looks like there's a flake in the integration test added.

```
=== RUN   TestLightningNetworkDaemon/connection_timeout
    network.go:87: 
        	Error Trace:	network.go:87
        	            				network.go:46
        	            				lnd_test.go:135
        	            				lnd_test.go:14482
        	Error:      	"rpc error: code = Unknown desc = server is still in the process of starting" does not contain "i/o timeout"
        	Test:       	TestLightningNetworkDaemon/connection_timeout
        	Messages:   	expected to get a timeout error, instead got: rpc error: code = Unknown desc = server is still in the process of starting
```

yyforyongyu commented:
> @yyforyongyu looks like there's a flake in the integration test added.
> 
> ```
> === RUN   TestLightningNetworkDaemon/connection_timeout
>     network.go:87: 
>         	Error Trace:	network.go:87
>         	            				network.go:46
>         	            				lnd_test.go:135
>         	            				lnd_test.go:14482
>         	Error:      	"rpc error: code = Unknown desc = server is still in the process of starting" does not contain "i/o timeout"
>         	Test:       	TestLightningNetworkDaemon/connection_timeout
>         	Messages:   	expected to get a timeout error, instead got: rpc error: code = Unknown desc = server is still in the process of starting
> ```

@wpaulino now fixed. The original call `node.ConnectPeer` is now wrapped into a new func `connect ` to allow additional waiting. And I've changed the git commits so it's easier to see the new changes.

wpaulino commented:
Needs another rebase.

yyforyongyu commented:
> Needs another rebase.

Rebased. However I got this weird lint error, which didn't happen in master, 
```
rpcserver.go:2000:4: singleCaseSwitch: should rewrite switch statement to if statement (gocritic)
			switch update := fundingUpdate.Update.(type) {
```
Should I send a PR to fix it or add a commit to fix it in current PR?

wpaulino commented:
Feel free to tack it on to this PR if it's a small change.

yyforyongyu commented:
> Feel free to tack it on to this PR if it's a small change.

A small change is added to fix the lint issue.
